### PR TITLE
PYTHON-1937 Add explicit encryption API

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Still todo: 
- [x] Create jira ticket
- [x] Add test for create_data_key.
- [x] Fix copy/pasted doc strings.

Jira: https://jira.mongodb.org/browse/PYTHON-1937
Patch: https://evergreen.mongodb.com/version/5d48d9c657e85a23e31c596f

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
Updated patch: https://evergreen.mongodb.com/version/5d49cf332a60ed73d4cef09d

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 1]
```diff
@@ -366,7 +388,130 @@ def __init__(self, ctx, command):
...
+          - `kms_provider`: The KMS provider.
+          - `opts`: An optional class:`DataKeyOpts`.
+        """
+        super(DataKeyContext, self).__init__(ctx)
+        try:
+            if kms_provider == 'aws':
+                if opts is None or opts.master_key is None:
+                    raise ValueError('opts.master_key is required for '
+                                     'kms_provider: "aws"')
+                if 'region' not in opts.master_key or 'key' not in \
```
Comment (behackett) [inline bindings/python/pymongocrypt/mongocrypt.py]: Use parens, not \ for line continuation.

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 2]
```diff
@@ -366,7 +388,130 @@ def __init__(self, ctx, command):
...
+                        opts.master_key:
+                    raise ValueError(
+                        'opts.master_key must include "region" and '
+                        '"key" for kms_provider: "aws"')
+                region = opts.master_key['region']
+                key = opts.master_key['key']
+                if PY3:
```
Comment (behackett) [inline bindings/python/pymongocrypt/mongocrypt.py]: Same question about unicode in python 2.

File: bindings/python/test/test_mongocrypt.py [hunk 3]
```diff
@@ -223,6 +225,9 @@ def fetch_keys(self, filter):
     def insert_data_key(self, data_key):
         raise NotImplementedError
 
+    def bson_encode(self, doc):
+        return BSON.encode(doc)
```
Comment (behackett) [inline bindings/python/test/test_mongocrypt.py]: Yuck. You have bson.encode now and none of this will work with older PyMongo versions.

File: bindings/python/test/test_mongocrypt.py [hunk 4]
```diff
@@ -261,6 +266,69 @@ def test_decrypt(self):
...
+    def __init__(self, kms_reply=None):
+        super(KeyVaultCallback, self).__init__(kms_reply=kms_reply)
+        self.data_key = None
+
+    def fetch_keys(self, filter):
+        return self.data_key
+
+    def insert_data_key(self, data_key):
+        self.data_key = data_key
+        return BSON(data_key).decode()['_id']
```
Comment (behackett) [inline bindings/python/test/test_mongocrypt.py]: bson.decode()

File: bindings/python/test/test_mongocrypt.py (hunk 3 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py]: I added the `bson_encode(self, doc) ` abstract method to deal with the fact that libmongocrypt requires the keyAltName to be passed as a BSON encoded document: https://jira.mongodb.org/browse/CDRIVER-3275. Instead of having pymongocrypt require pymongo I've implemented bson encoding via dependency injection.

This test implements a mock version of the callback API that pymongo will implement. In pymongo we can implement `bson_encode` using `bson.encode`.

> You have bson.encode now and none of this will work with older PyMongo versions.

I don't there's any concern for compatibility. This code uses the standard BSON.encode() API that has existed in pymongo forever. Only the tests for pymongocrypt require pymongo.

File: bindings/python/test/test_mongocrypt.py (hunk 4 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py]: We can't use `bson.decode()` because it hasn't been released yet.

File: bindings/python/pymongocrypt/explicit_encrypter.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        """
+        self.master_key = master_key
+        self.key_alt_names = key_alt_names
+
+
+class ExplicitEncrypter(object):
+    def __init__(self, callback, mongo_crypt_opts):
+        """Encrypts and decrypts BSON values.
+
+        This class is used by a driver to support for explicit encryption and
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/explicit_encrypter.py]: Done.

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 6]
```diff
@@ -366,7 +388,130 @@ def __init__(self, ctx, command):
...
+            if not lib.mongocrypt_ctx_setopt_algorithm(
+                    ctx, opts.algorithm, -1):
+                self._raise_from_status()
+
+            if opts.key_id is not None:
+                with MongoCryptBinaryIn(opts.key_id) as binary:
+                    if not lib.mongocrypt_ctx_setopt_key_id(ctx, binary.bin):
+                        self._raise_from_status()
+
+            if opts.key_alt_name is not None:
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/mongocrypt.py]: Right now we're depending on libmongocrypt to report the error if both options are set or any of the options are invalid.

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 1 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/mongocrypt.py]: Fixed.

File: bindings/python/test/test_mongocrypt.py (hunk 4 above)
Comment (behackett) [reply bindings/python/test/test_mongocrypt.py]: Oh, right, since this lives outside of PyMongo this would make testing more difficult at the moment,

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 7]
```diff
@@ -351,7 +363,17 @@ def __init__(self, ctx, database, command):
         """
         super(EncryptionContext, self).__init__(ctx)
         self.database = database
-        self._encrypt_init(database, command)
+        try:
+            with MongoCryptBinaryIn(command) as binary:
+                if PY3:
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/mongocrypt.py]: I fixed this by adding a `str_to_bytes` method. I also added a test with a unicode aws `master_key`.

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 2 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/mongocrypt.py]: I fixed this by adding a `str_to_bytes` method.
